// can-simple-dom/dist/cjs/simple-dom/dom.js
/*can-simple-dom@0.4.8#simple-dom/dom*/
'use strict';
Object.defineProperty(exports, '__esModule', { value: true });
exports.createDocument = exports.voidMap = exports.HTMLSerializer = exports.HTMLParser = exports.Event = exports.Document = exports.Element = exports.Node = undefined;
var _node = require('./document/node.js');
var _node2 = _interopRequireDefault(_node);
var _element = require('./document/element.js');
var _element2 = _interopRequireDefault(_element);
var _document = require('./document.js');
var _document2 = _interopRequireDefault(_document);
var _event = require('./event.js');
var _event2 = _interopRequireDefault(_event);
var _htmlParser = require('./html-parser.js');
var _htmlParser2 = _interopRequireDefault(_htmlParser);
var _htmlSerializer = require('./html-serializer.js');
var _htmlSerializer2 = _interopRequireDefault(_htmlSerializer);
var _voidMap = require('./void-map.js');
var _voidMap2 = _interopRequireDefault(_voidMap);
function _interopRequireDefault(obj) {
    return obj && obj.__esModule ? obj : { default: obj };
}
function createDocument(serializer, parser) {
    var doc = new _document2.default();
    doc.__serializer = serializer;
    doc.__parser = parser;
    return doc;
}
exports.Node = _node2.default;
exports.Element = _element2.default;
exports.Document = _document2.default;
exports.Event = _event2.default;
exports.HTMLParser = _htmlParser2.default;
exports.HTMLSerializer = _htmlSerializer2.default;
exports.voidMap = _voidMap2.default;
exports.createDocument = createDocument;
