// steal-css/test/css_paths/dist/bundles/main.js
/*[system-bundles-config]*/
System.bundles = {"bundles/main.css!":["folder/main.css!steal-css"]};
/*config.js*/
define('config.js', function (require, exports, module) {
    var steal = require('@steal');
    steal.config({
        ext: { css: 'steal-css' },
        paths: {
            'bootstrap/*': 'folder/bootstrap/*',
            'steal-css': '../../css.js',
            'helpers': '../helpers.js'
        }
    });
});
/*steal-css*/
define('steal-css', function (require, exports, module) {
    var loader = require('@loader');
    var steal = require('@steal');
    var isNode = typeof process === 'object' && {}.toString.call(process) === '[object process]';
    var importRegEx = /@import [^uU]['"]?([^'"\)]*)['"]?/g;
    var resourceRegEx = /url\(['"]?([^'"\)]*)['"]?\)/g;
    var waitSeconds = loader.cssOptions && loader.cssOptions.timeout ? parseInt(loader.cssOptions.timeout, 10) : 60;
    var noop = function () {
    };
    var onloadCss = function (link, cb) {
        var styleSheets = getDocument().styleSheets, i = styleSheets.length;
        while (i--) {
            if (styleSheets[i].href === link.href) {
                return cb();
            }
        }
        setTimeout(function () {
            onloadCss(link, cb);
        });
    };
    function getDocument() {
        if (typeof doneSsr !== 'undefined' && doneSsr.globalDocument) {
            return doneSsr.globalDocument;
        }
        return typeof document === 'undefined' ? undefined : document;
    }
    function getHead() {
        var doc = getDocument();
        var head = doc.head || doc.getElementsByTagName('head')[0];
        if (!head) {
            var docEl = doc.documentElement || doc;
            head = doc.createElement('head');
            docEl.insertBefore(head, docEl.firstChild);
        }
        return head;
    }
    function CSSModule(address, source) {
        this.address = address;
        this.source = source;
    }
    CSSModule.prototype = {
        injectLink: function () {
            if (this._loaded) {
                return this._loaded;
            }
            if (this.linkExists()) {
                this._loaded = Promise.resolve('');
                return this._loaded;
            }
            var doc = getDocument();
            var styleSheets = doc.styleSheets;
            var address = this.address;
            var link = this.link = doc.createElement('link');
            link.type = 'text/css';
            link.rel = 'stylesheet';
            link.href = this.address;
            this._loaded = new Promise(function (resolve, reject) {
                var timeout = setTimeout(function () {
                    reject('Unable to load CSS');
                }, waitSeconds * 1000);
                var loadCB = function (event) {
                    clearTimeout(timeout);
                    link.removeEventListener('load', loadCB);
                    link.removeEventListener('error', loadCB);
                    if (event && event.type === 'error') {
                        reject('Unable to load CSS');
                    } else {
                        resolve('');
                    }
                };
                if ('isApplicationInstalled' in navigator || !link.addEventListener) {
                    onloadCss(link, loadCB);
                } else {
                    link.addEventListener('load', loadCB);
                    link.addEventListener('error', loadCB);
                }
                getHead().appendChild(link);
            });
            return this._loaded;
        },
        injectStyle: function () {
            var doc = getDocument();
            var head = getHead();
            var style = this.style = doc.createElement('style');
            style.type = 'text/css';
            if (style.styleSheet) {
                style.styleSheet.cssText = this.source;
            } else {
                style.appendChild(doc.createTextNode(this.source));
            }
            head.appendChild(style);
        },
        updateUrls: function () {
            var rawSource = this.source, address = this.address;
            this.source = rawSource.replace(importRegEx, function (whole, part) {
                if (isNode) {
                    return '@import url(' + part + ')';
                } else {
                    return '@import url(' + steal.joinURIs(address, part) + ')';
                }
            });
            if (!isNode) {
                this.source = this.source + '/*# sourceURL=' + address + ' */';
                this.source = this.source.replace(resourceRegEx, function (whole, part) {
                    return 'url(' + steal.joinURIs(address, part) + ')';
                });
            }
            return this.source;
        },
        getExistingNode: function () {
            var doc = getDocument();
            var selector = '[href=\'' + this.address + '\']';
            return doc.querySelector && doc.querySelector(selector);
        },
        linkExists: function () {
            var styleSheets = getDocument().styleSheets;
            for (var i = 0; i < styleSheets.length; ++i) {
                if (this.address === styleSheets[i].href) {
                    return true;
                }
            }
            return false;
        },
        setupLiveReload: function (loader, name) {
            var head = getHead();
            if (loader.liveReloadInstalled) {
                var cssReload = loader['import']('live-reload', { name: module.id });
                Promise.resolve(cssReload).then(function (reload) {
                    loader['import'](name).then(function () {
                        reload.once(name, function () {
                            head.removeChild(css.style);
                        });
                    });
                });
            }
        }
    };
    if (loader.isEnv('production')) {
        exports.fetch = function (load) {
            var css = new CSSModule(load.address);
            return css.injectLink();
        };
    } else {
        exports.instantiate = function (load) {
            var loader = this;
            var css = new CSSModule(load.address, load.source);
            load.source = css.updateUrls();
            load.metadata.deps = [];
            load.metadata.format = 'css';
            load.metadata.execute = function () {
                if (getDocument()) {
                    css.injectStyle();
                    css.setupLiveReload(loader, load.name);
                }
                return loader.newModule({ source: css.source });
            };
        };
    }
    exports.CSSModule = CSSModule;
    exports.locateScheme = true;
    exports.buildType = 'css';
    exports.includeInBuild = true;
});
/*helpers*/
define('helpers', ['exports'], function (exports) {
    'use strict';
    Object.defineProperty(exports, '__esModule', { value: true });
    exports.default = {
        waitForCssRules: function waitForCssRules(styleNode, callback) {
            if (typeof callback !== 'function') {
                throw new Error('No callback provided');
            }
            var poller = setInterval(function () {
                var rulesLoaded = false;
                try {
                    styleNode.sheet.cssRules;
                    rulesLoaded = true;
                } catch (e) {
                }
                if (rulesLoaded) {
                    clearInterval(poller);
                    callback();
                }
            }, 10);
        }
    };
});
/*main*/
define('main', [
    'helpers',
    '@steal',
    './folder/main.css'
], function (_helpers, _steal) {
    'use strict';
    var _helpers2 = _interopRequireDefault(_helpers);
    var _steal2 = _interopRequireDefault(_steal);
    function _interopRequireDefault(obj) {
        return obj && obj.__esModule ? obj : { default: obj };
    }
    var testImage = function testImage(selector, cb) {
        var image = new Image();
        image.onload = function () {
            cb();
        };
        image.onerror = function () {
            cb(selector);
            QUnit.ok(false, 'image not loaded');
            QUnit.start();
            removeMyself();
        };
        image.src = $(selector).css('background-image').replace(/url\("?/, '').replace(/"?\)/, '');
    };
    if (_steal2.default.isEnv('production')) {
        if (typeof window !== 'undefined' && window.QUnit) {
            var btn = $('.btn.btn-danger');
            QUnit.equal(btn.css('display'), 'inline-block', '@import "locate://"; styles applied');
            QUnit.equal(btn.css('backgroundColor'), 'rgb(255, 0, 0)', '@import url("locate://"); styles applied');
            testImage('#test-element', function (err) {
                if (err) {
                    QUnit.ok(false, err);
                    QUnit.start();
                    removeMyself();
                } else {
                    QUnit.ok(true, 'background-image: url(../); styles applied');
                    testImage('#test-relative', function (err) {
                        if (err) {
                            QUnit.ok(false, err);
                        } else {
                            QUnit.ok(true, 'background-image: url(locate://); styles applied');
                        }
                        QUnit.start();
                        removeMyself();
                    });
                }
            });
        } else {
            console.log('background-image', $('#test-element').css('background-image'));
            console.log('tilde', $('#test-relative').css('background-image'));
        }
    } else {
        _helpers2.default.waitForCssRules($('style')[0], function () {
            if (typeof window !== 'undefined' && window.QUnit) {
                var btn = $('.btn.btn-danger');
                QUnit.equal(btn.css('display'), 'inline-block', '@import "locate://"; styles applied');
                QUnit.equal(btn.css('backgroundColor'), 'rgb(255, 0, 0)', '@import url("locate://"); styles applied');
                testImage('#test-element', function (err) {
                    if (err) {
                        QUnit.ok(false, err);
                        QUnit.start();
                        removeMyself();
                    } else {
                        QUnit.ok(true, 'background-image: url(../); styles applied');
                        testImage('#test-relative', function (err) {
                            if (err) {
                                QUnit.ok(false, err);
                            } else {
                                QUnit.ok(true, 'background-image: url(locate://); styles applied');
                            }
                            QUnit.start();
                            removeMyself();
                        });
                    }
                });
            } else {
                console.log('background-image', $('#test-element').css('background-image'));
                console.log('tilde', $('#test-relative').css('background-image'));
            }
        });
    }
});
